In [ ]:
import Pkg; Pkg.activate("/p/project1/netenesyquma/jung/statistical-CTMRG")
using statCTMRG
using PEPSKit
using TensorKit
using KrylovKit

  Activating project at `/p/project1/netenesyquma/jung/statistical-CTMRG`


## Test PEPKit's implementation

In [2]:
## AFM phase
a, b = 2, 1.7
Z = InfinitePartitionFunction(generate_local_tensor_Z2(a,b))
D = 50
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
env, = leading_boundary(env₀, Z; tol = 1.0e-8, maxiter = 7000);

[ Info: CTMRG init:	obj = -1.486096544616e+00	err = 1.0000e+00
┌ Warning: CTMRG cancel 7000:	obj = -5.153028632165e+02	err = 9.4894025508e-02	time = 20.38 min
└ @ PEPSKit ~/.julia/packages/PEPSKit/zz9BS/src/algorithms/ctmrg/ctmrg.jl:153


In [3]:
_, _, λ_h_1, λ_v_1 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(1))
_, _, λ_h_0, λ_v_0 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(0))
@show λ_h_1
@show λ_h_0
@show λ_v_1
@show λ_v_0
nothing

λ_h_1 = Vector{ComplexF64}[[0.9998739008749369 - 0.00028721468670985374im, 0.9606499386487696 + 0.2773077631752386im, -0.9606469458650567 - 0.27731528732152977im, -0.9998731099472886 + 0.00029527346857926684im, 0.8540091760990018 - 0.07711515914348213im, 0.7992092514565801 + 0.31071368334406096im, -0.7992003429615582 - 0.31069470554223255im, -0.8539953582480068 + 0.0770993925573053im, 0.7479836613457791 + 0.0007016713541216738im, 0.7188950184583123 + 0.20656718621007866im, -0.7182074047942871 - 0.20828717890879989im, -0.7477995200437347 + 0.0011414996676475493im]]
λ_h_0 = Vector{ComplexF64}[[1.0 - 1.669446939219372e-17im, 0.9606913448605838 + 0.2776186951737275im, -0.9606883494661634 - 0.2776262195381806im, -0.9999992112547476 + 8.060169296172002e-6im, 0.8541389630798681 - 0.07687953223587274im, 0.7992207137141335 + 0.31098244564320676im, -0.7992118094781899 - 0.3109634634254806im, -0.8541251390389478 + 0.07686376816596112im, 0.7480710532016918 + 0.0009161815512542286im, 0.718919735278

!!! Notice for the AFM phase: ┌ Warning: CTMRG cancel 7000:	obj = -5.153028632165e+02	err = 9.4894025508e-02	time = 20.38 min

In [8]:
## FM phase
a, b = 0.3, 0.5
Z = InfinitePartitionFunction(generate_local_tensor_Z2(a,b))
D = 50
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
env, = leading_boundary(env₀, Z; tol = 1.0e-8, maxiter = 7000);

[ Info: CTMRG init:	obj = +3.140690412895e-01	err = 1.0000e+00
[ Info: CTMRG conv 397:	obj = +1.011352464748e+00	err = 4.3764674681e-09	time = 1.07 min


In [9]:
_, _, λ_h_1, λ_v_1 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(1))
_, _, λ_h_0, λ_v_0 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(0))
@show λ_h_1
@show λ_h_0
@show λ_v_1
@show λ_v_0

nothing

λ_h_1 = Vector{ComplexF64}[[0.5891016329512255 - 0.0im, 0.5766771101709619 - 0.0im, 0.5563539131592337 - 0.0im, 0.5287252799653441 - 0.0im, 0.4946493905953842 - 0.0im, 0.45526516318327354 - 0.0im, 0.4119998310209204 - 0.0im, 0.36656420673477386 - 0.0im, 0.3209295463924485 - 0.0im, 0.27726550693105945 - 0.0im, 0.2377997133359775 - 0.0im, 0.20460055201193283 - 0.0im]]
λ_h_0 = Vector{ComplexF64}[[1.0 - 0.0im, 0.3400401927544824 - 0.0im, 0.32298021363560186 - 0.0im, 0.3224433010152782 - 0.0im, 0.30110017083664176 - 0.0im, 0.29951146923013705 - 0.0im, 0.29632255967561144 - 0.0im, 0.2729787915103176 - 0.0im, 0.2728372020113259 - 0.0im, 0.26879666047780876 - 0.0im, 0.26459595253540136 - 0.0im, 0.2429554159115461 - 0.0im, 0.2404761354207021 - 0.0im]]
λ_v_1 = Vector{ComplexF64}[[0.5891016329367993 - 0.0im, 0.576677110121282 - 0.0im, 0.5563539130643596 - 0.0im, 0.5287252798175965 - 0.0im, 0.4946493903945831 - 0.0im, 0.45526516294831193 - 0.0im, 0.4119998307799662 - 0.0im, 0.36656420650575167 - 0

In [10]:
## PM phase
a, b = 1.5, 0.5
Z = InfinitePartitionFunction(generate_local_tensor_Z2(a,b))
D = 50
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
env, = leading_boundary(env₀, Z; tol = 1.0e-8, maxiter = 7000);

[ Info: CTMRG init:	obj = -2.585772273511e+00	err = 1.0000e+00
[ Info: CTMRG conv 179:	obj = +3.333746987442e+00	err = 5.5898484634e-09	time = 28.96 sec


In [11]:
_, _, λ_h_1, λ_v_1 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(1))
_, _, λ_h_0, λ_v_0 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(0))
@show λ_h_1
@show λ_h_0
@show λ_v_1
@show λ_v_0

nothing

λ_h_1 = Vector{ComplexF64}[[1.0000000000000044 + 0.0im, -0.5232917648272672 + 0.0im, -0.5101732613347643 + 0.0im, -0.4887794913188117 + 0.0im, -0.45983757217448695 + 0.0im, -0.4244016363577832 + 0.0im, -0.38387321374440697 + 0.0im, -0.34000579500755734 + 0.0im, -0.29487969706914524 + 0.0im, 0.2663322535545791 + 0.0im, 0.25119024374536747 + 0.0im, -0.250833841173561 + 0.0im, 0.24749901291432896 + 0.0im]]
λ_h_0 = Vector{ComplexF64}[[1.0 + 0.0im, -0.5232917648272655 + 0.0im, -0.510173261334764 + 0.0im, -0.4887794913188121 + 0.0im, -0.45983757217448473 + 0.0im, -0.4244016363577855 + 0.0im, -0.38387321374440475 + 0.0im, -0.3400057950075584 + 0.0im, -0.29487969706914413 + 0.0im, 0.26633225355457885 + 0.0im, 0.25119024374536864 + 0.0im, -0.2508338411735605 + 0.0im, 0.24749901291432788 + 0.0im]]
λ_v_1 = Vector{ComplexF64}[[1.0000000000000067 - 0.0im, -0.5232917647615419 - 0.0im, -0.510173261088382 - 0.0im, -0.48877949083881267 - 0.0im, -0.45983757149453786 - 0.0im, -0.42440163555926896 - 0.0im

!!! Try again AFM phase for D=50 and D=60, (2,2) unit cell

In [29]:
## AFM phase
a, b = 2, 1.7
Z = InfinitePartitionFunction(fill(generate_local_tensor_Z2(a,b), 2, 2))
D = 60
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
ctmrg_alg = PEPSKit.CTMRGAlgorithm(;
                            maxiter = 7000,
                            verbosity = 4,
                            #trunc = truncerror(; atol = 1.0e-10)
                            trunc = truncrank(D)
                            )
env, = leading_boundary(env₀, Z, ctmrg_alg);

[ Info: CTMRG init:	obj = +1.814468444981e+04	err = 1.0000e+00
[ Info: CTMRG   1:	obj = -3.764701785342e+04	err = 3.5575924252e-01	time = 1.41 sec
[ Info: CTMRG   2:	obj = -3.531922423687e+04	err = 2.6009972883e-01	time = 0.22 sec
[ Info: CTMRG   3:	obj = +3.139932354062e+02	err = 2.0344479416e-01	time = 0.19 sec
[ Info: CTMRG   4:	obj = -4.004167793539e+03	err = 2.8423793163e-01	time = 0.19 sec
[ Info: CTMRG   5:	obj = -6.402024627635e+03	err = 2.9427371505e-01	time = 0.12 sec
[ Info: CTMRG   6:	obj = -8.929702897009e-01	err = 2.3896130471e-01	time = 0.19 sec
[ Info: CTMRG   7:	obj = -3.618226071213e+03	err = 3.2494657497e-01	time = 0.19 sec
[ Info: CTMRG   8:	obj = -1.965872117560e+03	err = 3.9270822737e-01	time = 0.21 sec
[ Info: CTMRG   9:	obj = -4.807453235832e+00	err = 3.7857323752e-01	time = 0.19 sec
[ Info: CTMRG  10:	obj = +1.472943220692e+03	err = 5.0828138009e-01	time = 0.12 sec
[ Info: CTMRG  11:	obj = -9.015443260358e+03	err = 4.4334929053e-01	time = 0.27 sec
[ Info: CTMRG

In [30]:
_, _, λ_h_1, λ_v_1 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(1))
_, _, λ_h_0, λ_v_0 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(0))
@show λ_h_1
@show λ_h_0
@show λ_v_1
@show λ_v_0

nothing

λ_h_1 = Vector{ComplexF64}[[0.5357628692171635 + 0.0im, 0.520386634070466 + 0.0im, 0.49536778169004686 + 0.0im, 0.4616333609953111 + 0.0im, 0.4204994829142522 + 0.0im, 0.37366523464986995 + 0.0im, 0.3231749155834772 + 0.0im, 0.27133415937242233 + 0.0im, 0.22057021198800597 + 0.0im, 0.1732354686022794 + 0.0im, 0.13137667696859448 + 0.0im, 0.12874610200012587 + 0.0im], [0.5357628690715863 + 0.0im, 0.5203866335182602 + 0.0im, 0.49536778056600633 + 0.0im, 0.4616333592680325 + 0.0im, 0.42049948066678333 + 0.0im, 0.3736652320508933 + 0.0im, 0.3231749128645907 + 0.0im, 0.2713341567753721 + 0.0im, 0.22057020970647026 + 0.0im, 0.173235466758377 + 0.0im, 0.1313766756130563 + 0.0im, 0.1287461019667616 + 0.0im]]
λ_h_0 = Vector{ComplexF64}[[1.0 + 0.0im, 0.2775971132936631 + 0.0im, 0.25829509908567244 + 0.0im, 0.2550158878936194 + 0.0im, 0.23120991201059093 + 0.0im, 0.23092471979665252 + 0.0im, 0.22580746384862846 + 0.0im, 0.20230976745132556 + 0.0im, 0.19971774002668422 + 0.0im, 0.1955181905064497 

In [2]:
## AFM phase
a, b = 2, 1.7
Z = InfinitePartitionFunction(fill(generate_local_tensor_Z2(a,b), 2, 2))
D = 50
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
ctmrg_alg = PEPSKit.CTMRGAlgorithm(;
                            maxiter = 7000,
                            verbosity = 4,
                            trunc = truncrank(D)
                            )
env, = leading_boundary(env₀, Z, ctmrg_alg);

[ Info: CTMRG init:	obj = +4.720581010442e+02	err = 1.0000e+00
[ Info: CTMRG   1:	obj = -9.990475194583e+06	err = 3.6246646941e-01	time = 9.13 sec
[ Info: CTMRG   2:	obj = -3.001980898998e+06	err = 2.7656310771e-01	time = 0.35 sec
[ Info: CTMRG   3:	obj = -1.105338298029e+04	err = 3.2127622813e-01	time = 0.09 sec
[ Info: CTMRG   4:	obj = +9.067629426787e+01	err = 4.3349711305e-01	time = 0.04 sec
[ Info: CTMRG   5:	obj = -5.900140104184e+02	err = 4.5379680643e-01	time = 0.10 sec
[ Info: CTMRG   6:	obj = +6.213057918424e+02	err = 4.2382777193e-01	time = 0.19 sec
[ Info: CTMRG   7:	obj = +2.830811461166e+03	err = 4.1192164236e-01	time = 0.10 sec
[ Info: CTMRG   8:	obj = -2.737890572141e+04	err = 4.5260004166e-01	time = 0.10 sec
[ Info: CTMRG   9:	obj = -9.414223188771e+03	err = 4.6277495448e-01	time = 0.10 sec
[ Info: CTMRG  10:	obj = -8.103338016690e+03	err = 5.6497628169e-01	time = 0.10 sec
[ Info: CTMRG  11:	obj = -3.444657484536e+05	err = 5.1674675368e-01	time = 0.10 sec
[ Info: CTMRG

In [3]:
_, _, λ_h_1, λ_v_1 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(1))
_, _, λ_h_0, λ_v_0 = PEPSKit._correlation_length(env; num_vals = 12, sector=Z2Irrep(0))
@show λ_h_1
@show λ_h_0
@show λ_v_1
@show λ_v_0

nothing

λ_h_1 = Vector{ComplexF64}[[0.5357628691953839 + 0.0im, 0.5203866339787092 + 0.0im, 0.4953677814922483 + 0.0im, 0.46163336068840916 + 0.0im, 0.4204994825114334 + 0.0im, 0.3736652341697307 + 0.0im, 0.3231749150605502 + 0.0im, 0.271334158854319 + 0.0im, 0.2205702115202033 + 0.0im, 0.17323546821647626 + 0.0im, 0.13137667667932407 + 0.0im, 0.12874610199617842 + 0.0im], [0.535762869021271 + 0.0im, 0.5203866333327865 + 0.0im, 0.4953677801880262 + 0.0im, 0.4616333586764641 + 0.0im, 0.4204994798978035 + 0.0im, 0.3736652311763455 + 0.0im, 0.3231749119496897 + 0.0im, 0.2713341558834382 + 0.0im, 0.2205702089085718 + 0.0im, 0.17323546610728305 + 0.0im, 0.13137667512949544 + 0.0im, 0.12874610195484587 + 0.0im]]
λ_h_0 = Vector{ComplexF64}[[1.0 + 0.0im, 0.27759711324293884 + 0.0im, 0.2582950990142976 + 0.0im, 0.2550158877538299 + 0.0im, 0.23120991183507916 + 0.0im, 0.23092471969188477 + 0.0im, 0.22580746364628682 + 0.0im, 0.20230976724446764 + 0.0im, 0.19971773980884036 + 0.0im, 0.19551819041300442 +

**Remark 1:** Convergence is very tricky for the AFM phase. Increasing the unit cell size and starting from various random initial states help. 

**Remark 2:** PEPSKit's CTMRG results give the expected degeneracy for the different phases FM : PM : AFM = 1 : 2 : 1

**Remark 3:** At the fixed point, we expect λ_h = λ_v

### Different truncation options for the projector

In [26]:
## FM phase
a, b = 0.3, 0.5
Z = InfinitePartitionFunction(generate_local_tensor_Z2(a,b))
D = 50
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
ctmrg_alg = PEPSKit.CTMRGAlgorithm(;
                            maxiter = 7000,
                            verbosity = 2,
                            trunc = truncspace(Venv)
                            )
env, = leading_boundary(env₀, Z, ctmrg_alg);

[ Info: CTMRG init:	obj = +5.743429937318e-01	err = 1.0000e+00
[ Info: CTMRG conv 105:	obj = +1.011352464748e+00	err = 9.8293077955e-09	time = 19.51 sec


In [27]:
@show space.(env.corners)
@show space.(env.edges)

nothing

space.(env.corners) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 1, 1}[Rep[ℤ₂](0 => 50, 1 => 50) ← Rep[ℤ₂](0 => 50, 1 => 50); Rep[ℤ₂](0 => 50, 1 => 50) ← Rep[ℤ₂](0 => 50, 1 => 50); Rep[ℤ₂](0 => 50, 1 => 50) ← Rep[ℤ₂](0 => 50, 1 => 50); Rep[ℤ₂](0 => 50, 1 => 50) ← Rep[ℤ₂](0 => 50, 1 => 50);;;]
space.(env.edges) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 2, 1}[(Rep[ℤ₂](0 => 50, 1 => 50) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 50, 1 => 50); (Rep[ℤ₂](0 => 50, 1 => 50) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 50, 1 => 50); (Rep[ℤ₂](0 => 50, 1 => 50) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 50, 1 => 50); (Rep[ℤ₂](0 => 50, 1 => 50) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 50, 1 => 50);;;]


In [28]:
## FM phase
a, b = 0.3, 0.5
Z = InfinitePartitionFunction(generate_local_tensor_Z2(a,b))
D = 50
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
ctmrg_alg = PEPSKit.CTMRGAlgorithm(;
                            maxiter = 7000,
                            verbosity = 2,
                            trunc = truncrank(D)
                            )
env, = leading_boundary(env₀, Z, ctmrg_alg);

[ Info: CTMRG init:	obj = -4.537679400431e-02	err = 1.0000e+00
[ Info: CTMRG conv 144:	obj = +1.011352464748e+00	err = 4.2690633704e-09	time = 3.67 sec


In [29]:
@show space.(env.corners)
@show space.(env.edges)

nothing

space.(env.corners) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 1, 1}[Rep[ℤ₂](0 => 24, 1 => 26) ← Rep[ℤ₂](0 => 24, 1 => 26); Rep[ℤ₂](0 => 24, 1 => 26) ← Rep[ℤ₂](0 => 24, 1 => 26); Rep[ℤ₂](0 => 24, 1 => 26) ← Rep[ℤ₂](0 => 24, 1 => 26); Rep[ℤ₂](0 => 24, 1 => 26) ← Rep[ℤ₂](0 => 24, 1 => 26);;;]
space.(env.edges) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 2, 1}[(Rep[ℤ₂](0 => 24, 1 => 26) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 24, 1 => 26); (Rep[ℤ₂](0 => 24, 1 => 26) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 24, 1 => 26); (Rep[ℤ₂](0 => 24, 1 => 26) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 24, 1 => 26); (Rep[ℤ₂](0 => 24, 1 => 26) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 24, 1 => 26);;;]


In [42]:
## FM phase
a, b = 0.3, 0.5
Z = InfinitePartitionFunction(generate_local_tensor_Z2(a,b))
D = 5
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
ctmrg_alg = PEPSKit.CTMRGAlgorithm(;
                            maxiter = 7000,
                            verbosity = 2,
                            trunc =  truncerror(; atol = 1.0e-15) # dynamically change virtual space s.t. the SVD truncation error is < atol
                            )
env, = leading_boundary(env₀, Z, ctmrg_alg);

[ Info: CTMRG init:	obj = -2.003702607453e-01	err = 1.0000e+00
[ Info: CTMRG conv 52:	obj = +1.011352464748e+00	err = 5.5358302622e-09	time = 7.19 sec


In [43]:
@show space.(env.corners)
@show space.(env.edges)

nothing

space.(env.corners) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 1, 1}[Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30); Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30); Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30); Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30);;;]
space.(env.edges) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 2, 1}[(Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 29, 1 => 30); (Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 29, 1 => 30); (Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 29, 1 => 30); (Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 29, 1 => 30);;;]


In [46]:
## FM phase
a, b = 0.3, 0.5
Z = InfinitePartitionFunction(generate_local_tensor_Z2(a,b))
D = 5
Venv = Z2Space(0 => D, 1 => D)
env₀ = CTMRGEnv(Z, Venv)
ctmrg_alg = PEPSKit.CTMRGAlgorithm(;
                            maxiter = 7000,
                            verbosity = 2,
                            trunc =  trunctol(; atol = 1.0e-15) # dynamically change virtual space s.t. the smallest SV is > atol
                            )
env, = leading_boundary(env₀, Z, ctmrg_alg);

[ Info: CTMRG init:	obj = -2.914860388101e-02	err = 1.0000e+00
[ Info: CTMRG conv 68:	obj = +1.011352464748e+00	err = 9.9186552078e-09	time = 11.30 sec


In [47]:
@show space.(env.corners)
@show space.(env.edges)

nothing

space.(env.corners) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 1, 1}[Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30); Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30); Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30); Rep[ℤ₂](0 => 29, 1 => 30) ← Rep[ℤ₂](0 => 29, 1 => 30);;;]
space.(env.edges) = TensorMapSpace{GradedSpace{Z2Irrep, Tuple{Int64, Int64}}, 2, 1}[(Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 29, 1 => 30); (Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)) ← Rep[ℤ₂](0 => 29, 1 => 30); (Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 29, 1 => 30); (Rep[ℤ₂](0 => 29, 1 => 30) ⊗ Rep[ℤ₂](0 => 1, 1 => 1)') ← Rep[ℤ₂](0 => 29, 1 => 30);;;]


## Compare to Erik's code

In [12]:
pattern_arr = Array{Any}(undef,1,1)
pattern_arr[1,1] = 1
D = 50
free_energy, env_arr = ctmrg([generate_local_tensor_Z2(a,b)], D, pattern_arr; maxiter = 7000, ϵ = 1e-8 , Space_type = :Z2, Projector_type = :half, 
                            conv_info = true, output_envs = true, svd_type = :envbond, adjust_χ = false, temp = 1, lognorm = 0, ftol = 10^-10,
                            reuse_envs = false);

CTM-RG iteration 1
the largest truncation error for the projectors is 1.8250120749944284e-8
the smallest truncation error for the projectors is 0.0
the relative difference of the real part of the free energies is 1.0
CTM-RG iteration 2
the largest truncation error for the projectors is 2.1073424255447017e-8
the smallest truncation error for the projectors is 0.0
the relative difference of the real part of the free energies is 0.18698470248839402
CTM-RG iteration 3
the largest truncation error for the projectors is 2.1073424255447017e-8
the smallest truncation error for the projectors is 0.0
largest SV-difference of the C tensors is unconverged
the relative difference of the real part of the free energies is 0.00560315615853964
CTM-RG iteration 4
the largest truncation error for the projectors is 2.356080457693621e-8
the smallest truncation error for the projectors is 1.4901161193847656e-8
largest SV-difference of the C tensors is unconverged
the relative difference of the real part of 

In [41]:
@show space(env_arr[1].ul)
@show space(env_arr[1].l)

space((env_arr[1]).ul) = Rep[ℤ₂](0 => 25, 1 => 25) ← Rep[ℤ₂](0 => 25, 1 => 25)
space((env_arr[1]).l) = Rep[ℤ₂](0 => 25, 1 => 25) ← (Rep[ℤ₂](0 => 1, 1 => 1) ⊗ Rep[ℤ₂](0 => 25, 1 => 25))


Rep[ℤ₂](0 => 25, 1 => 25) ← (Rep[ℤ₂](0 => 1, 1 => 1) ⊗ Rep[ℤ₂](0 => 25, 1 => 25))

In [ ]:
function tn_mul_4(X, env_arr)
    @tensor Y[-1, -2; -3] := env_arr.u[-1, 1, 2] * env_arr.d[-2, 3, 1] *
                             X[2, 3, -3]
    return Y
end


V_Z2_even = Z2Space(0 => 1, 1 => 0) 
X = TensorMap(randn,  space(env_arr[1].u, 1) ⊗ space(env_arr[1].d, 1),  V_Z2_even)
spectrum_even, _ = eigsolve(X, 12, :LM) do x return tn_mul_4(x, env_arr[1]) end

V_Z2_odd = Z2Space(0 => 0, 1 => 1) 
X = TensorMap(randn,  space(env_arr[1].u, 1) ⊗ space(env_arr[1].d, 1),  V_Z2_odd)
spectrum_odd, _ = eigsolve(X, 12, :LM) do x return tn_mul_4(x, env_arr[1]) end

V_Z2 = Z2Space(0 => 1, 1 => 1) 
X = TensorMap(randn,  space(env_arr[1].u, 1) ⊗ space(env_arr[1].d, 1),  V_Z2)
spectrum_all, _ = eigsolve(X, 13, :LM) do x return tn_mul_4(x, env_arr[1]) end

@show spectrum_even
@show spectrum_odd
@show spectrum_all

nothing

spectrum_even = ComplexF64[0.909611294187321 + 0.0im, -0.9096112941567214 + 0.0im, -0.6649884901669341 + 0.0im, 0.6649884901452559 + 0.0im, -0.6529337035876548 + 0.0im, 0.6529337035664173 + 0.0im, -0.6329071685914455 + 0.0im, 0.6329071685700004 + 0.0im, -0.6050545098404093 + 0.0im, 0.605054509819481 + 0.0im, -0.5696791926314744 + 0.0im, 0.5696791926127137 + 0.0im]
spectrum_odd = ComplexF64[0.9096112941873183 + 0.0im, -0.909611294156713 + 0.0im, -0.6649884901669391 + 0.0im, 0.6649884901452564 + 0.0im, -0.6529337035876607 + 0.0im, 0.6529337035664238 + 0.0im, -0.6329071685914409 + 0.0im, 0.6329071685700084 + 0.0im, -0.6050545098404115 + 0.0im, 0.6050545098194895 + 0.0im, -0.5696791926314799 + 0.0im, 0.5696791926127154 + 0.0im]
spectrum_all = ComplexF64[0.9096112941873163 + 0.0im, 0.909611294187315 + 0.0im, -0.9096112941567287 + 0.0im, -0.909611294156718 + 0.0im, -0.6649884901669368 + 0.0im, -0.6649884901669363 + 0.0im, 0.6649884901452549 + 0.0im, 0.6649884901452497 + 0.0im, -0.65293370358

In [48]:
pattern_arr = Array{Any}(undef,2,2)
pattern_arr[1,1] = 1
pattern_arr[1,2] = 1
pattern_arr[2,1] = 1
pattern_arr[2,2] = 1
D = 50
free_energy, env_arr = ctmrg([generate_local_tensor_Z2(a,b)], D, pattern_arr; maxiter = 7000, ϵ = 1e-8 , Space_type = :Z2, Projector_type = :half, 
                            conv_info = true, output_envs = true, svd_type = :envbond, adjust_χ = false, temp = 1, lognorm = 0, ftol = 10^-10,
                            reuse_envs = false);

CTM-RG iteration 1
the largest truncation error for the projectors is 0.0
the smallest truncation error for the projectors is 0.0
the relative difference of the real part of the free energies is 1.0
CTM-RG iteration 2
the largest truncation error for the projectors is 2.1073424255447017e-8
the smallest truncation error for the projectors is 0.0
the relative difference of the real part of the free energies is 0.07666297848439937
CTM-RG iteration 3
the largest truncation error for the projectors is 3.494632065164618e-8
the smallest truncation error for the projectors is 0.0
largest SV-difference of the C tensors is unconverged
the relative difference of the real part of the free energies is 0.006665340653153075
CTM-RG iteration 4
the largest truncation error for the projectors is 2.9802322387695312e-8
the smallest truncation error for the projectors is 0.0
largest SV-difference of the C tensors is unconverged
the relative difference of the real part of the free energies is 0.000617991249

In [49]:
V_Z2_even = Z2Space(0 => 1, 1 => 0) 
X = TensorMap(randn,  space(env_arr[1].u, 1) ⊗ space(env_arr[1].d, 1),  V_Z2_even)
spectrum_even, _ = eigsolve(X, 12, :LM) do x return tn_mul_4(x, env_arr[1]) end

V_Z2_odd = Z2Space(0 => 0, 1 => 1) 
X = TensorMap(randn,  space(env_arr[1].u, 1) ⊗ space(env_arr[1].d, 1),  V_Z2_odd)
spectrum_odd, _ = eigsolve(X, 12, :LM) do x return tn_mul_4(x, env_arr[1]) end

V_Z2 = Z2Space(0 => 1, 1 => 1) 
X = TensorMap(randn,  space(env_arr[1].u, 1) ⊗ space(env_arr[1].d, 1),  V_Z2)
spectrum_all, _ = eigsolve(X, 13, :LM) do x return tn_mul_4(x, env_arr[1]) end

@show spectrum_even
@show spectrum_odd
@show spectrum_all

nothing

spectrum_even = ComplexF64[1.0149438018981032 + 0.0im, 0.34773939995556197 + 0.0im, 0.33461680220093676 + 0.0im, 0.33351290448018023 + 0.0im, 0.31690236644926445 + 0.0im, 0.315446838466683 + 0.0im, 0.31347643872482 + 0.0im, 0.29483669478314634 + 0.0im, 0.29367406693966874 + 0.0im, 0.2897923507045884 + 0.0im, 0.2890605036337034 + 0.0im, 0.2697860035051765 + 0.0im, 0.267247903964718 + 0.0im]
spectrum_odd = ComplexF64[0.5988136430725777 + 0.0im, 0.5889274251269431 + 0.0im, 0.5727990137053914 + 0.0im, 0.5508754039604955 + 0.0im, 0.5237014662965717 + 0.0im, 0.4919251366272355 + 0.0im, 0.4563270401887374 + 0.0im, 0.4178474686778807 + 0.0im, 0.3775982377116322 + 0.0im, 0.33685870063015444 + 0.0im, 0.29705600333656923 + 0.0im, 0.25972526229126397 + 0.0im]
spectrum_all = ComplexF64[1.0149438018980996 + 0.0im, 0.5988136430725777 + 0.0im, 0.5889274251269417 + 0.0im, 0.5727990137053939 + 0.0im, 0.5508754039604986 + 0.0im, 0.5237014662965764 + 0.0im, 0.49192513662724136 + 0.0im, 0.45632704018873194

**Remark 4:** Increasing to (2,2) unit cell makes the AFM phase non-degenerate. There are two equivalent AFM configurations. In a (1,1) unit cell, CTMRG folds the 2 configs into each symmetry sector. This also breaks the translation symmetry.